src/api/demo.rkt

```diff
@@ -360,129 +390,118 @@
-                             (define formula (read-syntax 'web (open-input-string formula-str)))
-                             (define sample (hash-ref post-data 'sample))
-                             (define seed (hash-ref post-data 'seed #f))
-                             (define test (parse-test formula))
-                             (define pcontext (json->pcontext sample (test-context test)))
-                             (define command
-                               (create-job 'explanations
-                                           test
-                                           #:seed seed
-                                           #:pcontext pcontext
-                                           #:profile? #f
-                                           #:timeline-disabled? #t))
-                             (define id (start-job command))
-                             (wait-for-job id))))
-
-(define analyze-endpoint
+                             (define job-id (start-job (func post-data)))
+                             (hasheq 'job job-id 'path (make-path job-id)))))
+(define (make-sync-endpoint func)
   (post-with-json-response (lambda (post-data)
-                             (define formula-str (hash-ref post-data 'formula))
-                             (define formula (read-syntax 'web (open-input-string formula-str)))
-                             (define sample (hash-ref post-data 'sample))
-                             (define seed (hash-ref post-data 'seed #f))
-                             (define test (parse-test formula))
-                             (define pcontext (json->pcontext sample (test-context test)))
-                             (define command
-                               (create-job 'errors
-                                           test
-                                           #:seed seed
-                                           #:pcontext pcontext
-                                           #:profile? #f
-                                           #:timeline-disabled? #t))
-                             (define id (start-job command))
-                             (wait-for-job id))))
+                             (define job-id (start-job (func post-data)))
+                             (wait-for-job job-id))))
+
+; /api/sample endpoint: test in console on demo page:
+;; (await fetch('/api/sample', {method: 'POST', body: JSON.stringify({formula: "(FPCore (x) (- (sqrt (+ x 1))))", seed: 5})})).json()
+(define (sample-common post-data)
+  (define formula-str (hash-ref post-data 'formula))
+  (define formula (read-syntax 'web (open-input-string formula-str)))
+  (define seed* (hash-ref post-data 'seed))
+  (define test (parse-test formula))
+  (create-job 'sample test #:seed seed* #:pcontext #f #:profile? #f #:timeline-disabled? #t))
+
+(define sample-endpoint (make-sync-endpoint sample-common))
+(define start-sample-endpoint (make-async-endpoint sample-common))
```

Comment:
Definitely an improvement. One thing you can consider is something like this:

```
(define-syntax-rule (define-endpoint ([sync-name async-name] post-data) body ...)
  (begin
    (define (f post-data)
      body ...)
    (define sample-endpoint (make-sync-endpoint f))
    (define start-sample-endpoint (make-async-endpoint f))))
```

Reply:
This basically lets you write:

```
(define-endpoint ([sample-endpoint start-sample-endpoint] post-data)
  ... body of sample-common ...)
```

And it will put in the `make-xxx-endpoint` calls in for you and save you from having to name the function.